## Primero importaré pandas y exportaré la base de datos

In [3]:
import pandas as pd

df = pd.read_csv('https://breathecode.herokuapp.com/asset/internal-link?id=809&path=housing.csv')

df

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,MedHouseVal
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422
...,...,...,...,...,...,...,...,...,...
20635,1.5603,25.0,5.045455,1.133333,845.0,2.560606,39.48,-121.09,0.781
20636,2.5568,18.0,6.114035,1.315789,356.0,3.122807,39.49,-121.21,0.771
20637,1.7000,17.0,5.205543,1.120092,1007.0,2.325635,39.43,-121.22,0.923
20638,1.8672,18.0,5.329513,1.171920,741.0,2.123209,39.43,-121.32,0.847


## Debemos aplicar KMEANS sobre este conjunto de datos clasificandolo en 6 clusters

In [5]:
from sklearn.cluster import KMeans

model = KMeans(n_clusters=6, random_state=42)

model.fit(df)

df['cluster'] = model.predict(df)

df

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,MedHouseVal,cluster
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526,0
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585,1
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521,0
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413,0
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422,0
...,...,...,...,...,...,...,...,...,...,...
20635,1.5603,25.0,5.045455,1.133333,845.0,2.560606,39.48,-121.09,0.781,0
20636,2.5568,18.0,6.114035,1.315789,356.0,3.122807,39.49,-121.21,0.771,0
20637,1.7000,17.0,5.205543,1.120092,1007.0,2.325635,39.43,-121.22,0.923,0
20638,1.8672,18.0,5.329513,1.171920,741.0,2.123209,39.43,-121.32,0.847,0


## AHora voy a ver las características promedio de cada grupo

In [6]:
model.cluster_centers_

array([[ 3.94967976e+00,  3.17676756e+01,  5.72118953e+00,
         1.13812763e+00,  6.92393946e+02,  2.76528341e+00,
         3.59117335e+01, -1.19788243e+02,  2.12352940e+00],
       [ 3.75280241e+00,  2.30701695e+01,  5.08618568e+00,
         1.06598810e+00,  2.45594881e+03,  3.17909759e+00,
         3.52562441e+01, -1.19256054e+02,  2.00175822e+00],
       [ 3.80183604e+00,  2.86064324e+01,  5.23696217e+00,
         1.06347379e+00,  1.42779979e+03,  3.01067115e+00,
         3.55106458e+01, -1.19498051e+02,  2.03538916e+00],
       [ 4.49264637e+00,  1.26145251e+01,  5.72481919e+00,
         1.07432806e+00,  8.01038547e+03,  1.45807943e+01,
         3.50582123e+01, -1.18913408e+02,  2.11107838e+00],
       [ 2.44080000e+00,  1.25000000e+01,  5.31750761e+00,
         1.03508435e+00,  3.21240000e+04,  6.08944099e+00,
         3.49950000e+01, -1.19605000e+02,  1.26600000e+00],
       [ 3.95026978e+00,  1.82948403e+01,  5.21935904e+00,
         1.07280649e+00,  4.20842138e+03,  4.068360

## Los centroides permiten identificar las zonas geográficas y niveles de ingreso característicos de cada cluster, facilitando la interpretación de los patrones encontrados en los datos. Por lo que vamos a graficar esta información y analizar los componentes principales (PCA)


In [8]:
import matplotlib.pyplot as plt
from sklearn.decomposition import PCA

pca = PCA(n_components=3)

df_pca =pca.fit_transform(df.drop('cluster', axis=1))

df_pca = pd.DataFrame(df_pca, columns=['CP1', 'CP2', 'CP3'])

df_pca


,CP1,CP2,CP3
0,-1103.512710,8.589489,-0.782763
1,975.541206,-4.655319,-1.032619
2,-929.549944,20.047768,-1.975863
3,-867.550082,20.343342,-2.317450
4,-860.550444,20.340685,-2.697616
...,...,...,...
20635,-580.462778,-5.465242,0.497375
20636,-1069.436782,-13.965474,2.371430
20637,-418.437547,-12.918928,1.021984
20638,-684.439494,-12.806002,0.977709


## Ahora simplifique mi espacio en 3 dimensiones para mis datos, lo que me permitirá graficarlo posteriormente

In [9]:
centroides = pca.transform(model.cluster_centers_)

f:\ANACONDA\Lib\site-packages\sklearn\utils\validation.py:2749: UserWarning: X does not have valid feature names, but PCA was fitted with feature names
  warnings.warn(


## Procedemos a graficar nuevamente pero ahoira en 3 dimensiones

## Instalaré las librerías de Ploty y NBFormat

In [12]:
!pip install plotly

In [14]:
!pip install nbformat

In [15]:
import plotly.express as px

df_pca['cluster'] = df['cluster']
print(df_pca)
fig = px.scatter_3d(df_pca, x='CP1', y='CP2', z='CP3', color='cluster')

fig.show()

               CP1        CP2       CP3  cluster
0     -1103.512710   8.589489 -0.782763        0
1       975.541206  -4.655319 -1.032619        1
2      -929.549944  20.047768 -1.975863        0
3      -867.550082  20.343342 -2.317450        0
4      -860.550444  20.340685 -2.697616        0
...            ...        ...       ...      ...
20635  -580.462778  -5.465242  0.497375        0
20636 -1069.436782 -13.965474  2.371430        0
20637  -418.437547 -12.918928  1.021984        0
20638  -684.439494 -12.806002  0.977709        0
20639   -38.436218 -12.681171  1.045906        2

[20640 rows x 4 columns]


## Aqui se pueden ver los 6 clusters representados en 6 colores, de amarillo a morado.

## Ahora entrenaremos un modelo supervisado para predecir los clusters.

In [22]:
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

X = df.drop('cluster', axis=1)
y = df.cluster

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.8, random_state=42)

model_s = KNeighborsClassifier(n_neighbors=6)
model_s.fit(X_train, y_train)
y_pred = model_s.predict(X_test)

accuracy_score(y_test, y_pred)

0.999031007751938

## Al entrenar con las etiquetas obtenidas en KMEANS y luego con KNEIGHBORS, nuestra presición será del 99.90%, los cuales pueden trabajar conjuntamente porque KMEANS termina siendo la versión no supervisada de KNEIGHBORS, y tienen la misma lógica de puntos por distancia.

## Procedo a gardar el modelo

In [23]:
from pickle import dump

dump(model_s, open ("../models/KMEANS.model", "wb"))

## CONCLUSIÓN:

## Primero agrupe las casas, luego simplifique los datos para verlos, y finalmente entrene un modelo para clasificarlas automáticamente; es decir, un sistema que entiende como se divide el mercado a través de los clusters, una forma de ver el mercado de forma sencilla y con el KMEANS se puede clasificar cualquier casa nueva automáticamente, que me ayudaría a elegir la casa con un nivel de plusvalía más alto.